migrator/src/main/java/com/redhat/lightblue/migrator/ThreadMonitor.java

```diff
@@ -0,0 +1,279 @@
+            if(status==Status.alive) {
+                if(System.currentTimeMillis()-lastPing>threadTimeout) {
+                    kill();
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        /**
+         * Returns true if thread is killed before and now it is abandoned
+         */
+        public boolean abandon() {
+            if(status==Status.killed&&killTime>0) {
+                if(System.currentTimeMillis()-killTime>threadTimeout) {
+                    status=Status.abandoned;
+                    ((Thread)thread).interrupt();
+                    threadStatusChanged(thread);
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        @Override
+        public String toString() {
+            return "Thread="+thread+", lastPing:"+(new Date(lastPing).toString())+", status="+status+
+                " msg="+lastPingMsg+(killTime>0?", killlTime:"+(new Date(killTime)):"");
+        }
+        
+    }
+    
+    private final Map<MonitoredThread,ThreadStatus> threadMap=new HashMap<>();
+
+    public ThreadMonitor(long threadTimeout) {
+        this.threadTimeout=threadTimeout;
+    }
+
+    public ThreadMonitor() {}
+
+    /**
+     * Registers a thread status listener
+     */
+    public synchronized void registerThreadStatusListener(ThreadStatusListener listener) {
+        statusListeners.add(listener);
+    }
+
+    /**
+     * Unregisters a thread status listener
+     */
+    public synchronized void unregisterThreadStatusListener(ThreadStatusListener listener) {
+        statusListeners.remove(listener);
+    }
+
+    protected void threadStatusChanged(MonitoredThread thread) {
+        ArrayList<ThreadStatusListener> l;
+        synchronized(this) {
+            l=new ArrayList<>(statusListeners);
+        }
+        for(ThreadStatusListener x:l)
+            x.threadStatusChanged(thread,this);
+    }
+
+    public long getThreadTimeout() {
+        return threadTimeout;
+    }
+    
+    public void ping(String msg) {
+        Thread currentThread=Thread.currentThread();
+        LOGGER.debug("Ping from {}: {}",currentThread,msg);
+        if(currentThread instanceof MonitoredThread) {
+            MonitoredThread thread=(MonitoredThread)currentThread;
+            ThreadStatus status=threadMap.get(thread);
+            if(status==null) {
+                synchronized(this) {
+                    if(!threadMap.containsKey(thread)) {
+                        threadMap.put(thread,status=new ThreadStatus(thread));
+                    } else {
+                        status=threadMap.get(thread);
+                    }
+                }
+            }
+            status.lastPingMsg=msg;
+            LOGGER.debug("Status:{}",status);
+            long now=System.currentTimeMillis();
+            if(status.status==Status.killed) {
+                ((Thread)thread).interrupt();
```

Comment:
Done

On Wed, Apr 27, 2016 at 9:05 AM, Dennis Crissman <notifications@github.com>
wrote:

> In migrator/src/main/java/com/redhat/lightblue/migrator/ThreadMonitor.java
> <https://github.com/lightblue-platform/lightblue-migrator/pull/336#discussion_r61275444>
> :
>
> > +            MonitoredThread thread=(MonitoredThread)currentThread;
> > +            ThreadStatus status=threadMap.get(thread);
> > +            if(status==null) {
> > +                synchronized(this) {
> > +                    if(!threadMap.containsKey(thread)) {
> > +                        threadMap.put(thread,status=new ThreadStatus(thread));
> > +                    } else {
> > +                        status=threadMap.get(thread);
> > +                    }
> > +                }
> > +            }
> > +            status.lastPingMsg=msg;
> > +            LOGGER.debug("Status:{}",status);
> > +            long now=System.currentTimeMillis();
> > +            if(status.status==Status.killed) {
> > +                ((Thread)thread).interrupt();
>
> Could just use currentThread here to eliminate some casting.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/lightblue-platform/lightblue-migrator/pull/336/files/bd3352f8ca30985d5aca299f70c88f9d7ab44f92#r61275444>
>